src/python/pants/backend/jvm/tasks/jvm_compile/rsc/rsc_compile.py

```diff
@@ -480,242 +379,88 @@ def is_java_compile_target(t):
-          self.runtime_classpath_product.add_for_target(*args, **kwargs)
-          self.rsc_classpath_product.add_for_target(*args, **kwargs)
-
-      zinc_key = self._compile_against_rsc_key_for_target(compile_target)
-      zinc_jobs.append(
-        Job(
-          zinc_key,
-          functools.partial(
-            self._default_work_for_vts,
-            ivts,
-            compile_context_pair[1],
-            'runtime_classpath',
-            counter,
-            compile_contexts,
-            CompositeProductAdder(
-              runtime_classpath_product,
-              self.context.products.get_data('rsc_classpath'))),
-          [
-            self._compile_against_rsc_key_for_target(target)
-            for target in invalid_dependencies_without_jar_metacps],
-          self._size_estimator(compile_context_pair[1].sources),
-          on_failure=ivts.force_invalidate,
-        )
+    #   generate jars that make javac happy at this point.
+
+    def only_zinc_invalid_dep_keys(invalid_deps):
+      for tgt in invalid_deps:
+        if self._classify_compile_target(tgt) is not None:
+          yield self._zinc_key_for_target(tgt)
+
+    # NB: zinc jobs for rsc-compatible targets never depend on their own corresponding rsc jobs,
+    # just the rsc jobs of their dependencies!
+    # In order to implement sub-target compiles, that will need to change.
+    def make_zinc_job(target, input_product_key, output_products, dep_keys):
+      return Job(key=self._zinc_key_for_target(target),
+        fn=functools.partial(
+          self._default_work_for_vts,
+          ivts,
+          zinc_compile_context,
+          input_product_key,
+          counter,
+          compile_contexts,
+          CompositeProductAdder(*output_products)),
+        dependencies=list(dep_keys),
+        size=self._size_estimator(zinc_compile_context.sources),
+        on_success=ivts.update,
+        on_failure=ivts.force_invalidate,
       )
 
-      metacp_key = self._metacp_key_for_target(compile_target)
-      rsc_jobs.append(
-        Job(
-          metacp_key,
-          functools.partial(
-            work_for_vts_metacp,
-            ivts,
-            compile_context_pair[0],
-            'runtime_classpath'),
-            [self._metacp_dep_key_for_target(target) for target in invalid_dependencies] + [
-              'metacp(jdk)',
-              zinc_key,
-            ],
-          self._size_estimator(compile_context_pair[0].sources),
-          on_success=ivts.update,
-          on_failure=ivts.force_invalidate,
-        )
-      )
+    self._classify_compile_target(compile_target).resolve_for_enum_variant({
+      'zinc-only': lambda: zinc_jobs.append(
+        make_zinc_job(compile_target,
+          input_product_key='runtime_classpath',
+          output_products=[
+            runtime_classpath_product,
+            self.context.products.get_data('rsc_classpath')],
+          dep_keys=only_zinc_invalid_dep_keys(invalid_dependencies))),
```

Comment:
This ends up silently losing rsc parallelism for all the `zinc-only` scala targets (which is going to be the vast majority of targets), which is why I was thinking splitting the enum up into `-java` and `-scala` versions was a good idea. `zinc-only` targets still only depend on the rsc compiles of their dependencies -- except for targets with any java sources, since javac can't yet depend on rsc output.

As per https://github.com/pantsbuild/pants/pull/7227/files#diff-074dd871ff62c835ab90d4a0a1264a63R314 or https://github.com/pantsbuild/pants/pull/7227/files#diff-074dd871ff62c835ab90d4a0a1264a63R549: if we're electing to use this coarser strategy to avoid introducing the intermediate `zinc_scala_classpath_from_rsc` product in this diff, I would want an extremely clear TODO describing:
- what parallelism we're dropping (all non-rsc targets are waiting for the zinc compiles of all their dependencies).
- what changes to the dependency graph must be made in followups.

The end result of the mixed compile changes in terms of the execution graph and intermediate products was documented in several comments in #7227 and I'm concerned about losing that and mysteriously losing parallelism later if the performance implications of the dependency graph of the zinc jobs isn't documented here.

This is of course the most complex part of the whole thing and will be changing a lot, and imho this would become significantly more maintainable if we spent a small amount of time converting to v2 rules and letting the v2 engine construct the graph implicitly.

Reply:
As far as this PR is concerned, though, zinc-only targets == targets with java, right? So is this comment something to address when that stops being the case, rather than speculatively here?

I'm slightly confused, though, about the javac part of this comment, and how it relates to this line of code, which could maybe do with a clarifying code comment. As far as I know, we never invoke javac (we use zinc to compile java), and zinc is always happy with jars output by rsc. With that understanding, we _could_ use the rsc produced output on the classpath here. So could you add a comment explaining which part of my understanding is false, and why we need to use the output from zinc?

Reply:
> As far as this PR is concerned, though, zinc-only targets == targets with java, right? So is this comment something to address when that stops being the case, rather than speculatively here?

I did not realize this, see:

> for all the `zinc-only` scala targets (which is going to be the vast majority of targets)

so this review comment can largely be ignored. I would love at least to have a comment here describing why we are setting up the dependencies like we are right now (explaining that rsc cannot produce jars to satisfy javac yet).

> As far as I know, we never invoke javac (we use zinc to compile java), and zinc is always happy with jars output by rsc.

I was under the impression zinc invoked javac because it was consistently mentioned that producing jars to satisfy javac was an rsc TODO. If that was only a concern for warp because it doesn't use zinc, then my misunderstanding here is almost definitely the reason for any performance regression in the previous mixed compile branch.

Reply:
> This ends up silently losing rsc parallelism for all the `zinc-only` scala targets (which is going to be the vast majority of targets), which is why I was thinking splitting the enum up into `-java` and `-scala` versions was a good idea. 

Since this review doesn't include mixed compile support, zinc only compiles of targets with just scala that are not tests isn't supported, so I'm not sure this applies.

> `zinc-only` targets still only depend on the rsc compiles of their dependencies -- except for targets with any java sources, since javac can't yet depend on rsc output.

I think you're saying that the `zinc-only` workflow should use rsc output of their dependencies, apart from java source having targets. I didn't do that here, because I was trying to keep this change to just removing metacp as much as possible. I'd like to have adding additional workflows be separate patches so that their effects on behavior are clearer.

The current workflows are
- rsc then zinc: eg the current rsc-compatible scala flow. Where rsc jobs depend on rsc and zinc jobs depend on the rsc of the current jobs dependencies
- zinc only: each target is compiled by zinc and requires zinc artifacts on the classpath. Currently this applies to java containing targets and tests.

When we add mixed compile support, we could add another workflow
- zinc with rsc dependencies: for scala targets that fail to compile with rsc, but can build against rsc generated jars.


> I was under the impression zinc invoked javac
Yes. It does. Zinc is just a wrapper for Javac and Scalac. It adds some behavior on top, in particular doing analysis and supporting incremental builds using that.